graders/python/python_autograder/code_feedback.py

```diff
@@ -299,6 +299,112 @@ def bad(msg):
+
+        - ``name``: Name of the dictionary that is being checked. This will be used to give feedback.
+        - ``ref``: Reference dictionary.
+        - ``data``: Student dictionary to be checked.
+        - ``partial_keys``: If not None, it takes a List of keys to check if these particular keys are present in the student's dict or not.
+        - ``check_only_keys``: If true, grading will be done only based on checking all keys in student's dict and reference's dict match or not.
+        - ``check_only_values``: If true, grading will be done only based on checking all values in student's dict and reference's dict match or not.
+        - ``entry_type_key``: If not None, requires that each key in the student's dictionary in solution be of this type.
+        - ``entry_type_value``: If not None, requires that each value in the student's dictionary in solution be of this type.
+        - ``accuracy_critical``: If true, grading will halt on failure.
+        - ``report_failure``: If true, feedback will be given on failure.
+        """
+
+        def bad(msg):
+            if report_failure:
+                cls.add_feedback(msg)
+            if accuracy_critical:
+                cls.finish("")
+            else:
+                return False
+
+        if data is None:
+            return bad(f"{name} is None or not defined")
+
+        if not isinstance(data, dict):
+            return bad(f"{name} is not a dict")
+
+        if partial_keys is not None and len(partial_keys) >= 1:
+            for partial_key in partial_keys:
+                if partial_key not in data:
+                    return bad(f"{name} does not contain key {partial_key}")
+            return True
+
+        if partial_keys is None:
+            if check_only_values:
+                if len(ref.values()) != len(data.values()):
+                    return bad(
+                        f"{name} has the wrong length for values--expected {len(ref.values())}, got {len(data.values())}"
+                    )
+
+            if len(ref) != len(data):  # this is default length of keys check
+                return bad(
+                    f"{name} has the wrong number of entries, expected {len(ref)}, got {len(data)}"
+                )
+
+        if entry_type_value is not None:
+            for value in data.values():
+                if not isinstance(value, entry_type_value):
+                    return bad(
+                        f"{name} has the wrong type for value {value}, expecting type {entry_type_value}"
+                    )
+
+        if entry_type_key is not None:
+            for key in data.keys():
+                if not isinstance(key, entry_type_key):
+                    return bad(
+                        f"{name} has the wrong type for key {key}, expecting type {entry_type_key}"
+                    )
+
+        if check_only_keys or check_only_values:
+            check_keys = False
```

Comment:
I'm trying to tell what's going on here. Would this be better named `keys_valid`, `keys_checked`, `checked_keys`, or something like that? Or am I misunderstanding what this variable represents?

Reply:
I think these are confusing variable names. More expressive would be `check_keys` and `check_values`, defaulted to True. Then `check_keys` would require having `check_values` set to True. Let me think about this some more.

Reply:
Let me look into all the mentioned comments right now. And I will clarify them shortly.

Reply:
Lines 379 - 399

`check_only_keys` - When True, the grading will only consider if student's dict and reference's dict have the same `keys` or not. Once this comparison is made, the function has to return True if matched, False otherwise. 

`check_only_values` - When True, the grading will only consider if student's dict and reference's dict have the same `values` or not. Once this comparison is made, the function has to return True if matched, False otherwise. 

As you can see, `check_only_keys` and `check_only_values` are independent checks meaning as a user, I can just want to enable grading based on only values and choose to give credit, or based on only keys and just give credit, or choose to have both. In case, both are enabled then the logic should confirm that both these checks return True, and only then check_dict() can return true. ( This is handled in lines: 396, 397)

At the same time, if only one of these is enabled, they logic should be intact and return True if the specific check returns true.  ( This is handled in lines: 399)

And the `check_keys` and `check_values` are local flags helping us to build the above logic.


Is this what you are looking for? Please let me know your thoughts.

Reply:
Actually, I think I understand your question now. 

A better naming would be to use: 
`check_keys` instead of `check_only_keys` 
`check_values` instead of `check_only_values`  
`keys_valid` instead of `check_keys` 
`values_valid` instead of `check_values` 

Is this what you are suggesting?

Reply:
^I think this simpler naming would be helpful.

Reply:
Latest push includes these renamed variables 👍🏽